File: histomicstk/ComplementStainMatrix.py
import numpy as np


def ComplementStainMatrix(W):
    """Generates a complemented stain matrix
    Used to fill out empty columns of a stain matrix for use with
    ColorDeconvolution. Replaces right-most column with normalized
    cross-product of first two columns.

    Parameters
    ----------
    W : array_like
        A 3x3 stain calibration matrix with stain color vectors in columns.

    Returns
    -------
    WComp : array_like
        A 3x3 complemented stain calibration matrix with a third
        orthogonal column.

    See Also
    --------
    ColorDeconvolution
    """

    # copy input to output for initialization
    WComp = W

    # calculate directed cross-product of first two columns
    if (W[0, 0]**2 + W[0, 1]**2) > 1:
        WComp[0, 2] = 0
    else:
        WComp[0, 2] = (1 - (W[0, 0]**2 + W[0, 1]**2))**0.5

    if (W[1, 0]**2 + W[1, 1]**2) > 1:
        WComp[1, 2] = 0
    else:
        WComp[1, 2] = (1 - (W[1, 0]**2 + W[1, 1]**2))**0.5

    if (W[2, 0]**2 + W[2, 1]**2) > 1:
        WComp[2, 2] = 0
    else:
        WComp[2, 2] = (1 - (W[2, 0]**2 + W[2, 1]**2))**0.5

    # normalize new vector to unit-norm
    WComp[:, 2] = WComp[:, 2] / np.linalg.norm(WComp[:, 2])

    return WComp

File: macenko-pca/rust/complement_stain_matrix.rs
use crate::float_trait::MacenkoFloat;
use ndarray::{Array1, Array2, ArrayView1};
use ndarray_linalg::Norm;

/// Generates a complemented stain matrix.
/// Replaces right-most column with normalized cross-product of first two columns.
pub fn complement_stain_matrix<F: MacenkoFloat>(w: &Array2<F>) -> Array2<F> {
    let stain0 = w.column(0);
    let stain1 = w.column(1);
    let stain2 = cross_product(&stain0, &stain1);
    let stain2_l2 = stain2.norm_l2();
    let stain2_norm = stain2.mapv(|x| x / stain2_l2);
    let mut result = Array2::<F>::zeros((3, 3));
    result.column_mut(0).assign(&stain0);
    result.column_mut(1).assign(&stain1);
    result.column_mut(2).assign(&stain2_norm);
    result
}

/// Compute cross product of two 3-element vectors
fn cross_product<F: MacenkoFloat>(a: &ArrayView1<F>, b: &ArrayView1<F>) -> Array1<F> {
    Array1::from_vec(vec![
        a[1] * b[2] - a[2] * b[1],
        a[2] * b[0] - a[0] * b[2],
        a[0] * b[1] - a[1] * b[0],
    ])
}
